### ISSUE TITLE
[REVIEW]: tesliper: a theoretical spectroscopist's little helper

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4164]
**Submitting author:** @mishioo (<a href="http://orcid.org/0000-0001-7884-8982">Michał M. Więcław</a>)
**Repository:** https://github.com/mishioo/tesliper
**Branch with paper.md** (empty if default branch): 
**Version:** 0.9.3
**Editor:** @rkurchin
**Reviewers:** @alejandrogallo, @arepstein
**Archive:** 10.5281/zenodo.6448370

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/3dedbebd87078d759ac2c40afaf47138/status.svg)](https://joss.theoj.org/papers/3dedbebd87078d759ac2c40afaf47138)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/3dedbebd87078d759ac2c40afaf47138"><img src="https://joss.theoj.org/papers/3dedbebd87078d759ac2c40afaf47138/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/3dedbebd87078d759ac2c40afaf47138/status.svg)](https://joss.theoj.org/papers/3dedbebd87078d759ac2c40afaf47138)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/mishioo/tesliper" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@alejandrogallo & @arepstein, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @rkurchin know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @alejandrogallo
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/mishioo/tesliper">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/mishioo/tesliper">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@mishioo) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @arepstein
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/mishioo/tesliper">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/mishioo/tesliper">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@mishioo) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @alejandrogallo, @arepstein it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
PDF failed to compile for issue #4164 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.21 s (453.4 files/s, 112910.8 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          82           3135           4856          12979
reStructuredText                12            724            786           1384
Markdown                         1             44              0            199
YAML                             2              8              0             39
-------------------------------------------------------------------------------
SUM:                            97           3911           5642          14601
-------------------------------------------------------------------------------


Statistical information for the repository '8c65f255a2e7e1a8f58965ac' was
gathered on 2022/02/14.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
M. M. Więcław                   98          3826           4561            9.02
Michal M. Wieclaw               12           277            229            0.54
Michał M. Więcław              774         53120          31004           90.44

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Michał M. Więcław         20970           39.5         12.1                4.77
```

**Comment by rkurchin** [comment]
👋 reviewers @alejandrogallo and @arepstein, please let me know if you have any questions about the review process or how to get started!

**Comment by rkurchin** [comment]
Hi again @alejandrogallo and @arepstein, just a reminder to get your reviews started!

**Comment by alejandrogallo** [comment]
Hi @rkurchin, is it ok for reviewers to issue pull-requests for things like typos or should we open issues for these?

**Comment by rkurchin** [comment]
You are welcome to do either/both! Just make sure to follow any contribution guidelines issues by package authors. ;)

**Comment by rkurchin** [comment]
Looks like some review-related changes have been merged in – reminder to reviewers @alejandrogallo and @arepstein to take a look again if you haven't started already!

(Also, it's not required but helpful that if you open issues/PR's in the repo under review that you mention this issue for tracking purposes – helps me as the editor easily see that there's activity and not accidentally pester people when they've been active!)

**Comment by mishioo** [comment]
That's right, I've addressed (sufficiently, I hope) all issues pointed out so far by @alejandrogallo, and I was just wondering if I should mention this here. Thanks for having your finger on the pulse, @rkurchin!

**Comment by arepstein** [comment]
@rkurchin I don't think the paper compiled correctly. Is that an issue with the repo that I should raise or is it with JOSS? Thanks

**Comment by mishioo** [comment]
The paper is on a separate branch "joss", but maybe I'll just merge it to "master" to make its compilation straightforward.

**Comment by mishioo** [comment]
@whedon generate pdf

**Comment by mishioo** [comment]
@editorialbot generate pdf

**Comment by rkurchin** [comment]
Glad you worked this out! For future reference, we can compile papers from other branches, I just need to tell the bot which branch it should look on. So if you want to move it back, that's fine – just let me know.

**Comment by arepstein** [comment]
Thanks @mishioo !

**Comment by mishioo** [comment]
Oh, I was thinking about merging the paper after it's accepted anyway, so it's fine. Hopefully it will stay there for good. ;)

**Comment by arepstein** [comment]
Hi @rkurchin and @mishioo, I've completed my checklist and I think Tesliper meets all of the criteria for publication in JOSS. The scientific paper establishes a clear need for a free, open-source spectroscopy tool that includes conformers and it explains why Tesliper addresses that need. I tested basic functionality with some of my own Gaussian output files both using the GUI and a Python notebook and I found the documentation to be  clear and easy to follow. 

I have a few small comments. I can put them in issues if that would be more helpful, but they're not quite issues so I thought I'd write them here first.

- Some errors in Gaussian output files cause errors in glassware but not the parser. It might be nice to include warnings for bad log files or at least a log file lacking "Normal termination."
- I think the recursive parsing functionality is important, as running each conformer in its own directory is common, but I don't think that's available in the GUI. 
- I would consider adding test coverage guidelines for contributors, such as with [coveralls.io](https://coveralls.io/).
- In my work, I've had some issues with RMSD calculations for different algorithms and have often needed to try a few to see which work best for a given molecule. I would consider adding some additional options in addition to Kabsch. [Pymatgen's Molecule Matcher](https://github.com/materialsproject/pymatgen/blob/c37ba66cce99e93557b5e44d697059e2c93c0869/pymatgen/analysis/molecule_matcher.py) has a few that I would consider adding. The InchiMolAtomMapper is particularly nice when it works because I believe the atomic ordering does not need to be maintained between the conformers.

Thank you for writing this, it seems like it will be quite helpful to a wide range of research!

**Comment by mishioo** [comment]
Thank you for your insight, @arepstein! And also for words of appreciation, as I was a bit concerned if this project will be useful enough for a broader public, but I really hope so! Please, let me address you comments below:

- About errors: if you mean just adding a `logging.info` or a `logging.warning` at the end of parsing an output file that has an error or that ended abruptly, then it's easily doable. There is already a `logging.info` when parsing unexpectedly stops, but maybe changing it to `warning` would be better. Anyway, I'm not quite sure about errors in glassware part; when does they happen? If you feel this doesn't suppose to happen, then I would appreciate if you'd open an issue with more details, and, ideally, with an example log file.
- About recursive parsing: I'm glad you mentioned this. the group that have done initial testing didn't need this feature, but I was not sure about other groups workflow. I'll add it to GUI.
- About coverage: valid point, I'll update relevant READEME section.
- About RMDS: that's a tougher one... I am aware of the limitations of Kabsch/RMSD approach, especially when rotation of symmetrical groups like -Me or -Ph comes into play. I was thinking about expanding the conformers comparison functionality with other methods myself, but on the other hand, I feel like adding a Pymatgen, Open Babel, and SciPy as dependencies would be quite an overkill for such a humble project as `tesliper`... I think it's a valid route of development, but I'd prefer to leave it for now and come back to it at a later time, after some more research on my part. Is this acceptable?

**Comment by arepstein** [comment]
- For the errors, I do think a `warning` could be more helpful for a user debugging. But I would also leave that up to you, I don't think it's an absolutely necessary change. The example I used was deleting several frequencies as if the job timed out and that's when I got an error in the glassware -- it is a very expected error because the output file was wrong, so I think the error was supposed to happen and don't think an issue is necessary.
- For the RMSDs, I completely understand about not wanting to add large dependencies and agree that it makes sense as a future addition and is not necessary for publication.

**Comment by rkurchin** [comment]
Hi all, checking in on this – in particular, @alejandrogallo, do you have questions/issues?

And @arepstein, I know there are potentially a few things to discuss here regarding the abnormal termination warnings, but I see your checklist is done, so are you happy to accept this at its current state?

**Comment by arepstein** [comment]
Hi @rkurchin , yes I'm happy to accept at its current state!

**Comment by mishioo** [comment]
So happy to see the first green light for `tesliper`'s publication! :)
Also, an update from me, as I was finally able to introduce changes we talked about. @arepstein could you please confirm if they satisfy your suggestions?

- I've added a note about [installation in zsh](https://tesliper.readthedocs.io/en/latest/installation.html#install-from-terminal), as we discussed in [this issue](https://github.com/mishioo/tesliper/issues/5).
- I've added the recursive extraction to GUI & updated [the relevant part of docs](https://tesliper.readthedocs.io/en/latest/gui.html).
- I've updated testing guidelines in [relevant part of READEME](https://github.com/mishioo/tesliper#participating-in-code) with coverage info.
- I've resolved the matter of abnormal termination warnings by moving it up by one level (from `GaussianParser` to `Soxhlet` class) -- now a warning will be issued if any abnormally terminated files are detected in a single batch of extraction (i.e. only once for `extract()` call). I think this is a sane middle ground. Relevant changes [in this commit](https://github.com/mishioo/tesliper/commit/36159b7e7329b293c19b342226f7df7cb476b1ee).

I didn't issue a new release yet, I decided to wait with this for possible further comments from @alejandrogallo.

**Comment by arepstein** [comment]
That all looks great @mishioo !

**Comment by rkurchin** [comment]
Great, thanks for these updates. Pinging @alejandrogallo again for further feedback/questions!

**Comment by rkurchin** [comment]
👋 @alejandrogallo, checking in again!

**Comment by alejandrogallo** [comment]
I am sorry for the delay, I'll be finishing my review today

**Comment by alejandrogallo** [comment]
I have concluded my review of the most recent version. I have to thank also @arepstein for raising quite relevant points like the recursive parsing.

I am satisfied with the current status of the package (da2b096db6adf66e64e6fe0923f40f434d64c976) and **I recommend this package for publication at JOSS**.

The author @mishioo has done a great job addressing the points raised and further improving the documentation, the CI infrastructure, and thus the overall quality. I have to praise also the clear writing of the introductions with pertinent references to related software that I could therefore compare it to. I think `tesliper` fills a very relevant gap in the spectroscopy ecosystem and can now serve as a robust starting point for other projects.

Furthermore, the current state of the documentation for typical workflows, GUI program, scripting and advanced usage ensures that even if shortcomings appear in the future, the maintenance of the software is realistic and thus it represents a relevant long-term addition to the community.

The documentation is also clearly written for material scientists or quantum chemists in general as myself that are somewhat external to the field and need guidelines to get started quickly. Thus, the presented examples are well written, can be followed and work as intended in my testing.

Upon further reviewing of the code, I think that extensions to [`pymatgen`](https://pymatgen.org/), [`babel`](https://www.openbabel.org/), [`ase`](http://wiki.fysik.dtu.dk/ase) or related software packages will probably happen in the future and they are realistically implemented, but I concur with @arepstein that these are not necessary extensions for the current publication and can be left for future enhancements. Indeed, the clear structuring of the core package enables it.

I thank the author @mishioo for his outstanding work and the editor @rkurchin for her free software and open science commitment and excellent editorial task.

**Comment by mishioo** [comment]
Thank you @alejandrogallo and @arepstein that you have found time for this review. I'm very happy to see `tesliper` recommended for publication!

The commit mentioned by @alejandrogallo is the latest release, issued in the meantime with version number 0.9.3.
I have created a Zenodo archive for it, the assigned DOI is [10.5281/zenodo.6448370](https://doi.org/10.5281/zenodo.6448370).
Is there anything else I should do now, @rkurchin?

**Comment by rkurchin** [comment]
Haha, I was just writing a comment asking for the version tag and DOI, so you beat me to it on that one!

I'll do an editorial pass over the manuscript and send any comments shortly, and then we'll be ready to roll! 🚀

**Comment by rkurchin** [comment]
@editorialbot set 0.9.3 as version

**Comment by rkurchin** [comment]
@editorialbot set 10.5281/zenodo.6448370 as DOI

**Comment by rkurchin** [comment]
@editorialbot set 10.5281/zenodo.6448370 as archive

**Comment by rkurchin** [comment]
@editorialbot generate pdf

**Comment by rkurchin** [comment]
A few minor editorial suggestions:
* line 34: singular/plural disagreement – perhaps rephrase to "...adoption of one or more of the software packages mentioned above..." or "...adoption of one or more piece of the aforementioned software..."
* line 35: parallelism - suggest change to "...incomplete, or unable to be done in an automated fasion."
* 37: hyphenate "time-consuming"

**Comment by rkurchin** [comment]
@editorialbot check references

**Comment by mishioo** [comment]
Thank you very much for your suggestions, @rkurchin! I have introduced these changes to the paper.

**Comment by rkurchin** [comment]
@editorialbot recommend-accept

**Comment by arfon** [comment]
@editorialbot accept

**Comment by arfon** [comment]
@alejandrogallo, @arepstein – many thanks for your reviews here and to @rkurchin for editing this submission! JOSS relies upon the volunteer effort of people like you and we simply wouldn't be able to do this without you ✨ 

@mishioo – your paper is now accepted and published in JOSS :zap::rocket::boom:

**Comment by mishioo** [comment]
@rkurchin, @alejandrogallo, @arepstein, @arfon, thank you all for conducting this review, I'm extremely happy to see my software published in JOSS! :)
